### PR TITLE
Support encryption at rest for DynamoDB tables

### PR DESCRIPTION
**Pull request by ewbankkit**
Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/3298.
Added an `encrypt_at_rest` block to `aws_dynamodb_table` resource and data source, similar to `aws_elasticsearch_domain`.

Acceptance tests:
```
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSDynamoDbTable_basic -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_basic
--- PASS: TestAccAWSDynamoDbTable_basic (199.52s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	199.527s

make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_encryption'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSDynamoDbTable_encryption -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_encryption
--- PASS: TestAccAWSDynamoDbTable_encryption (55.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	55.729s

make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsDynamoDbTable_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccDataSourceAwsDynamoDbTable_basic -timeout 120m
=== RUN   TestAccDataSourceAwsDynamoDbTable_basic
--- PASS: TestAccDataSourceAwsDynamoDbTable_basic (49.30s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	49.304s

make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSDynamoDbTable_basic -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_basic
--- PASS: TestAccAWSDynamoDbTable_basic (199.52s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	199.527s
```

### CODE REVIEW

**Comment by ctreatma** [comment]
What more needs to be done in order to get this in to a release?

**Comment by bflad** [comment]
I'll try to give this a proper review this week.

File: aws/data_source_aws_dynamodb_table.go [hunk 1]
```diff
@@ -167,6 +167,20 @@ func dataSourceAwsDynamoDbTable() *schema.Resource {
 				Type:     schema.TypeInt,
 				Computed: true,
 			},
+			"encrypt_at_rest": {
```
Comment (bflad) [inline aws/data_source_aws_dynamodb_table.go]: 😅  Sorry for this, but could you rename all these `encrypt_at_rest` references to `server_side_encryption` so they better match the DynamoDB API `SSE`?

File: aws/resource_aws_dynamodb_table.go [hunk 2]
```diff
@@ -250,6 +265,16 @@ func resourceAwsDynamoDbTableCreate(d *schema.ResourceData, meta interface{}) er
 		}
 	}
 
+	if v, ok := d.GetOk("encrypt_at_rest"); ok {
+		options := v.([]interface{})
+		if options[0] == nil {
```
Comment (bflad) [inline aws/resource_aws_dynamodb_table.go]: I think this needs to check `len(options) == 0`

```go
s := make([]interface{}, 0)
if s[0] == nil {
	log.Println("All is well")
}

# panic: runtime error: index out of range

s := make([]interface{}, 0)
if len(s) == 0 {
	log.Println("All is well")
}

# 2018/03/01 15:19:00 All is well
```

Comment (bflad) [reply aws/resource_aws_dynamodb_table.go]: Generally speaking though, we forego error checking against our own schema.

File: aws/resource_aws_dynamodb_table_test.go [hunk 3]
```diff
@@ -625,6 +625,27 @@ func testAccCheckDynamoDbTableTimeToLiveWasUpdated(n string) resource.TestCheckF
...
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSDynamoDbTableDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSDynamoDbConfigInitialStateWithEncryption(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckInitialAWSDynamoDbTableExists("aws_dynamodb_table.basic-dynamodb-table", &conf),
+					testAccCheckInitialAWSDynamoDbTableConf("aws_dynamodb_table.basic-dynamodb-table", true),
+				),
```
Comment (bflad) [inline aws/resource_aws_dynamodb_table_test.go]: Can you please check the Terraform state here?

```go
resource.TestCheckResourceAttr("aws_dynamodb_table.basic-dynamodb-table", "server_side_encryption.#", "1"),
resource.TestCheckResourceAttr("aws_dynamodb_table.basic-dynamodb-table", "server_side_encryption.0.enabled", "true"),
```

File: aws/resource_aws_dynamodb_table_test.go [hunk 4]
```diff
@@ -927,6 +958,57 @@ resource "aws_dynamodb_table" "basic-dynamodb-table" {
 `, rName)
 }
 
+func testAccAWSDynamoDbConfigInitialStateWithEncryption(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_dynamodb_table" "basic-dynamodb-table" {
+  name = "%s"
+  read_capacity = 10
```
Comment (bflad) [inline aws/resource_aws_dynamodb_table_test.go]: Can you please set the test capacities to `1`? 💸

File: aws/resource_aws_dynamodb_table_test.go [hunk 5]
```diff
@@ -927,6 +958,57 @@ resource "aws_dynamodb_table" "basic-dynamodb-table" {
...
+    name = "TestTableHashKey"
+    type = "S"
+  }
+
+  attribute {
+    name = "TestTableRangeKey"
+    type = "S"
+  }
+
+  attribute {
```
Comment (bflad) [inline aws/resource_aws_dynamodb_table_test.go]: Can you please remove extraneous attributes/indexes here? We only really care about the encryption setting.

File: aws/resource_aws_dynamodb_table_test.go [hunk 6]
```diff
@@ -927,6 +958,57 @@ resource "aws_dynamodb_table" "basic-dynamodb-table" {
...
+  global_secondary_index {
+    name = "InitialTestTableGSI"
+    hash_key = "TestTableHashKey"
+    range_key = "TestGSIRangeKey"
+    write_capacity = 10
+    read_capacity = 10
+    projection_type = "KEYS_ONLY"
+	}
+
+	encrypt_at_rest {
```
Comment (bflad) [inline aws/resource_aws_dynamodb_table_test.go]: Nitpick: Indentation got thrown off here

File: aws/resource_aws_dynamodb_table_test.go (hunk 3 above)
Comment (ewbankkit) [reply aws/resource_aws_dynamodb_table_test.go]: Will do

File: aws/data_source_aws_dynamodb_table.go (hunk 1 above)
Comment (ewbankkit) [reply aws/data_source_aws_dynamodb_table.go]: Will do

File: aws/resource_aws_dynamodb_table_test.go (hunk 6 above)
Comment (ewbankkit) [reply aws/resource_aws_dynamodb_table_test.go]: Will do, copy and past flub.

File: aws/resource_aws_dynamodb_table_test.go (hunk 4 above)
Comment (ewbankkit) [reply aws/resource_aws_dynamodb_table_test.go]: Will do

File: aws/resource_aws_dynamodb_table_test.go (hunk 5 above)
Comment (ewbankkit) [reply aws/resource_aws_dynamodb_table_test.go]: Will do

File: aws/resource_aws_dynamodb_table_test.go (hunk 4 above)
Comment (ewbankkit) [reply aws/resource_aws_dynamodb_table_test.go]: Will do

File: aws/resource_aws_dynamodb_table_test.go (hunk 3 above)
Comment (ewbankkit) [reply aws/resource_aws_dynamodb_table_test.go]: By doing this and [removing the extraneous attributes](https://github.com/terraform-providers/terraform-provider-aws/pull/3303/files#r171683831) I can now revert the modifications to `testAccCheckInitialAWSDynamoDbTableConf` so as not to need the `sse bool` parameter

File: aws/resource_aws_dynamodb_table.go (hunk 2 above)
Comment (ewbankkit) [reply aws/resource_aws_dynamodb_table.go]: Doesn't `GetOk()` guard against `len(options) == 0`?

**Comment by ewbankkit** [comment]
Re-ran acceptance tests:

```
make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsDynamoDbTable_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccDataSourceAwsDynamoDbTable_basic -timeout 120m
=== RUN   TestAccDataSourceAwsDynamoDbTable_basic
--- PASS: TestAccDataSourceAwsDynamoDbTable_basic (63.78s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	63.792s
```

```
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_basic'==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSDynamoDbTable_basic -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_basic
--- PASS: TestAccAWSDynamoDbTable_basic (232.25s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	232.265s
```

```
make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSDynamoDbTable_encryption'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -run=TestAccAWSDynamoDbTable_encryption -timeout 120m
=== RUN   TestAccAWSDynamoDbTable_encryption
--- PASS: TestAccAWSDynamoDbTable_encryption (27.52s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	27.535s
```

**Comment by ewbankkit** [comment]
@bflad I have made the changes you suggested (question about the `len(options) == 0` comment above).
I have noticed that if I create a table

```
resource "aws_dynamodb_table" "basic-dynamodb-table" {
  name = "test-enc-0"
  read_capacity = 1
  write_capacity = 1
  hash_key = "TestTableHashKey"

  attribute {
    name = "TestTableHashKey"
    type = "S"
  }
}
```

and then change the code to

```
resource "aws_dynamodb_table" "basic-dynamodb-table" {
  name = "test-enc-0"
  read_capacity = 1
  write_capacity = 1
  hash_key = "TestTableHashKey"

  attribute {
    name = "TestTableHashKey"
    type = "S"
  }

  server_side_encryption {
    enabled = false
  }
}
```

`terraform plan` shows a diff

```
-/+ aws_dynamodb_table.basic-dynamodb-table (new resource required)
      id:                               "test-enc-0" => <computed> (forces new resource)
      arn:                              "arn:aws:dynamodb:us-west-2:000000000000:table/test-enc-0" => <computed>
      attribute.#:                      "1" => "1"
      attribute.2990477658.name:        "TestTableHashKey" => "TestTableHashKey"
      attribute.2990477658.type:        "S" => "S"
      hash_key:                         "TestTableHashKey" => "TestTableHashKey"
      name:                             "test-enc-0" => "test-enc-0"
      read_capacity:                    "1" => "1"
      server_side_encryption.#:         "0" => "1"
      server_side_encryption.0.enabled: "" => "false" (forces new resource)
      stream_arn:                       "" => <computed>
      stream_label:                     "" => <computed>
      stream_view_type:                 "" => <computed>
      write_capacity:                   "1" => "1"
```

**Comment by ewbankkit** [comment]
Ended up using the `CustomizeDiff` functionality to do this as `DiffSuppressFunc` doesn't support lists/maps.

**Comment by bflad** [comment]
@ewbankkit sounds like we should add an acceptance test for that case so its not made a regression in the future. I was able to get the existing testing in this PR passing with `Optional: true` and `Default: false` on the `enabled` attribute (removing the `CustomizeDiff`) but I did not test your new scenario.

Maybe we should rework the test suite slightly:
* Create a `func testAccAWSDynamoDbConfig_basic(rName string) string {` that's basically a duplicate of the new `testAccAWSDynamoDbConfigInitialStateWithEncryption` without the SSE config block
* Add `enabled bool` parameter to `testAccAWSDynamoDbConfigInitialStateWithEncryption` and 
 use `enabled = %t`
* Update `TestAccAWSDynamoDbTable_encryption` (or create a new separate test) to include new scenario you mentioned as two test steps
* Rename existing `TestAccAWSDynamoDbTable_basic` into something like `TestAccAWSDynamoDbTable_extended` (ugh naming is hard, but its just testing way too many optional configurations right now)
* Recreate `TestAccAWSDynamoDbTable_basic` at the new `testAccAWSDynamoDbConfig_basic` and check all the basic attributes in the state

If you don't have time to work on the above, definitely an above and beyond ask, please let us know and we can add commits after yours. 👍

**Comment by ewbankkit** [comment]
@bflad I'll make those suggested changes to the acceptance tests; Should be able to get to them today.

**Comment by ewbankkit** [comment]
@bflad I made those changes to the `aws_dynamodb_table` resource acceptance tests, using the names you suggested.
In `TestAccAWSDynamoDbTable_encryption` I use the table creation timestamp to make sure that the table gets recreated when the `enabled` value changes and doesn't get recreated when it shouldn't.

**Comment by bflad** [comment]
This has been released in [version 1.11.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1110-march-09-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**